### PR TITLE
Fix for other-authentication in GLPI 9.5

### PR DESCRIPTION
**Pull request by finalbeta**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #8455 #6324

### CODE REVIEW

**Review by cedric-anne — APPROVED** [review]

File: inc/auth.class.php [hunk 1]
```diff
@@ -919,6 +919,9 @@ function login($login_name, $login_password, $noauto = false, $remember_me = fal
                // Then ensure addslashes
                $input = Toolbox::addslashes_deep($input);
                unset ($this->user->fields);
+               if($authtype == self::EXTERNAL && !isset($input["authtype"])){
```
Comment (cedric-anne) [inline inc/auth.class.php:922]:
```suggestion
               if ($authtype == self::EXTERNAL && !isset($input["authtype"])) {
```

**Review by trasher — COMMENTED** [review]
I have no idea for this one :/

**Review by orthagh — APPROVED** [review]

**Comment by orthagh** [comment]
check all code parts of different auth mode

in case of connection with:
- glpidb
- ldap
- authmail

authtype is forced for adding or updating user

in case of sso login, we should never update a user (ex ldap account with sso login)
but in the case of addition on non existing user, the pr use the authtype and it seems a good case.

I think we can force the authtype without condition in the case of addition (and so the remove the if)

File: inc/auth.class.php [hunk 2]
```diff
@@ -919,6 +919,9 @@ function login($login_name, $login_password, $noauto = false, $remember_me = fal
                // Then ensure addslashes
                $input = Toolbox::addslashes_deep($input);
                unset ($this->user->fields);
+               if ($authtype == self::EXTERNAL && !isset($input["authtype"])) {
+                   $input["authtype"] = $authtype;
+               }
```
Comment (cedric-anne) [inline inc/auth.class.php:924]:
```suggestion
                $input["authtype"] = $authtype;
```

File: inc/auth.class.php [hunk 3]
```diff
@@ -919,6 +919,7 @@ function login($login_name, $login_password, $noauto = false, $remember_me = fal
                // Then ensure addslashes
                $input = Toolbox::addslashes_deep($input);
                unset ($this->user->fields);
+                $input["authtype"] = $authtype;
```
Comment (cedric-anne) [inline inc/auth.class.php:922]:
```suggestion
               $input["authtype"] = $authtype;
```

**Comment by cedric-anne** [comment]
> I think we can force the authtype without condition in the case of addition (and so the remove the if)

Removing it breaks LDAP test suite :/

File: inc/auth.class.php [hunk 4]
```diff
@@ -919,6 +919,7 @@ function login($login_name, $login_password, $noauto = false, $remember_me = fal
                // Then ensure addslashes
                $input = Toolbox::addslashes_deep($input);
                unset ($this->user->fields);
+               $input["authtype"] = $authtype;
```
Comment (cedric-anne) [inline inc/auth.class.php:922]:
```suggestion
               if ($authtype == self::EXTERNAL && !isset($input["authtype"])) {
                  $input["authtype"] = $authtype;
               }
```

**Review by trasher — APPROVED** [review]

**Pull request merged**